src/tarray.jl

```diff
@@ -195,3 +101,193 @@ function tfill(val::Real, dim)
     task_local_storage(res.ref, (n,d))
     return res
 end
+
+#
+# Conversion between TArray and Array
+#
+function Base.get(x::TArray)
+    n, d = task_local_storage(x.ref)
+    return d
+end
+
+_get(x) = x
+_get(x::TArray) = get(x)
+
+function Base.convert(::Type{Array}, x::TArray)
+    return convert(Array{eltype(x), ndims(x)}, x)
+end
+function Base.convert(::Type{Array{T,N}}, x::TArray{T,N}) where {T,N}
+    c = convert(Array{T, N}, deepcopy(get(x)))
+    return c
+end
+
+function Base.convert(::Type{TArray}, x::AbstractArray)
+    return convert(TArray{eltype(x),ndims(x)}, x)
+end
+function Base.convert(::Type{TArray{T,N}}, x::AbstractArray{T,N}) where {T,N}
+    res = TArray{T,N}()
+    n   = n_copies()
+    task_local_storage(res.ref, (n,x))
+    return res
+end
+
+#
+# Representation
+#
+function Base.display(x::TArray)
+    arr = x.orig_task.storage[x.ref][2]
+    @warn "display(::TArray) prints the originating task's storage, " *
+        "not the current task's storage. " *
+        "Please use show(::TArray) to display the current task's version of a TArray."
+    display(arr)
+end
+
+Base.show(io::IO, x::TArray) = Base.show(io::IO, task_local_storage(x.ref)[2])
+
+function Base.summary(io::IO, x::TArray)
+  print(io, "Task Local Array: ")
+  summary(io, get(x))
+end
+
+Base.print_array(io::IO, x::TArray) = Base.print_array(io, get(x))
+
+#
+# Iterator Interface
+#
+
+IteratorSize(::Type{TArray{T, N}}) where {T, N} = HasShape{N}()
+IteratorEltype(::Type{TArray}) = HasEltype()
+
+#
+# Forward many methods to the underlying array
+#
+for F in (:iterate, :eltype, :length, :size,
+          :firstindex, :lastindex, :ndims, :axes,
+          :collect)
+    @eval Base.$F(a::TArray, args...) = $F(get(a), args...)
+end
+
+#
+# Similarity implementation
+#
+
+Base.similar(x::TArray) = tzeros(eltype(x), size(x))
+Base.similar(x::TArray, ::Type{T}) where {T} = tzeros(T, size(x))
+Base.similar(x::TArray, dims::Dims) = tzeros(eltype(x), dims)
```

Comment:
As mentioned above, one should just define
```suggestion
Base.similar(x::TArray, ::Type{T}, dims::Dims) = TArray(similar(get(x), T, dims))
```

BTW in my opinion `get` is really a misuse of `Base.get` which is intended to retrieve a value from a collection or instead a default value if the key does not exist. It would be better to rename it and use an internal function.

Reply:
- you are again right about the function `similar`
- and it is a good point about the method name `get`, I will change it to `_get`